### PR TITLE
Add LlamaIndex Memory module

### PR DESCRIPTION
**Pull request by jerryjliu**
### Background
LlamaIndex is a tool that provides a data interface between LLM's and data. It contains a set of data structures that allow users to retrieve data from their knowledge corpus, in novel ways, for use with different LLM tasks (question-answering, summarization, etc.).

Note that some of our data structures can be backed by a vector store like Pinecone, Weaviate, Milvus, etc.) 

It makes sense to have LlamaIndex (especially the vector indexes) be a memory module for AutoGPT. 

### Changes
Added a LlamaIndexMemory wrapper. 

### Documentation
see LlamaIndex Setup section in README

### Test Plan
added llamaindex_memory_test.py

### PR Quality Checklist
- [x] My pull request is atomic and focuses on a single change.
- [x] I have thoroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes 

I think I checked off all boxes but please let me know any feedback!

### CODE REVIEW

File: autogpt/memory/llama_index.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
+"""LlamaIndex memory storage provider."""
+from llama_index.indices.vector_store.base import GPTVectorStoreIndex
```
Comment (richbeales) [inline autogpt/memory/llama_index.py]: Does this require a package from pip?  If so should be in requirements.txt

Comment (jerryjliu) [reply autogpt/memory/llama_index.py]: Just checking, I'd put the llama-index import in requirements.txt? 

Taking a look at `memory/__init__.py` as an example, it seems like imports for different memory modules are lazily loaded. E.g. for Milvus, pymilvus isn't in requirments.txt. 

Want to double-check if lazy import is ok or you'd want to make llama-index a required import

Comment (BillSchumacher) [reply autogpt/memory/llama_index.py]: Yeah we should not add this to requirements.

**Review by BillSchumacher — DISMISSED** [review]

**Comment by jerryjliu** [comment]
thanks @BillSchumacher - just resolved some readme merge conflicts - let me know if this looks good!

**Review by nponeccop — APPROVED** [review]

**Comment by jerryjliu** [comment]
@nponeccop thanks, just fixed lint errors (oops didn't install pre-commit hooks beforehand).

hope everything works now!! 🤞

File: autogpt/memory/__init__.py [hunk 2]
```diff
@@ -34,6 +34,14 @@
     MilvusMemory = None
 
 
+try:
+    from autogpt.memory.llama_index import LlamaIndexMemory
+    supported_memory.append("llama_index")
+except ImportError:
+    print("llama-index not installed. Skipping import.")
```
Comment (richbeales) [inline autogpt/memory/__init__.py]: please comment this print line, it has been removed from the other imports as people complained it was an error

Comment (jerryjliu) [reply autogpt/memory/__init__.py]: done!

File: README.md [hunk 3]
```diff
@@ -44,43 +44,6 @@ Development of this free, open-source project is made possible by all the <a hre
...
-  - [🧠 Memory pre-seeding](#-memory-pre-seeding)
-  - [💀 Continuous Mode ⚠️](#-continuous-mode-️)
-  - [GPT3.5 ONLY Mode](#gpt35-only-mode)
-  - [🖼 Image Generation](#-image-generation)
-  - [⚠️ Limitations](#️-limitations)
-  - [🛡 Disclaimer](#-disclaimer)
-  - [🐦 Connect with Us on Twitter](#-connect-with-us-on-twitter)
-  - [Run tests](#run-tests)
-  - [Run linter](#run-linter)
-
```
Comment (nponeccop) [inline README.md]: You deleted some parts of Readme during the merging by mistake

File: README.md [hunk 4]
```diff
@@ -389,7 +353,7 @@ in various ways.
 
 Querying an index/graph will both retrieve the set of relevant
 data from the knowledge corpus as well as (optionally) synthesize
-the final answer.
```
Comment (nponeccop) [inline README.md]: Revert unrelated whitespace fixes

**Review by nponeccop — CHANGES_REQUESTED** [review]

**Comment by nponeccop** [comment]
@jerryjliu And CI is red now

File: README.md (hunk 3 above)
Comment (jerryjliu) [reply README.md]: @nponeccop you might be on an outdated version of the README, the diffs don't show up on the right

**Comment by jerryjliu** [comment]
@nponeccop fixed the CI checks, seems like linting rules have changed since the last commit

**Review by nponeccop — APPROVED** [review]

**Comment by jerryjliu** [comment]
@nponeccop @richbeales @BillSchumacher apologies for the delay! conflicts should be resolved now, would love if you had time to take a look + approve workflow. thanks so much

**Comment by p-i-** [comment]
This is a mass message from the AutoGPT core team.
Our apologies for the ongoing delay in processing PRs.
This is because we are re-architecting the AutoGPT core!

For more details (and for infor on joining our Discord), please refer to:
  https://github.com/Significant-Gravitas/Auto-GPT/wiki/Architecting

**Comment by ntindle** [comment]
Hey, I've marked this as don't merge until the Memory Fixes are in. Sorry to keep it on hold longer, just no real way to test/validate functionality until the fixes are in

**Comment by Swiftyos** [comment]
Closing old PRS

**Pull request closed without merging**